### DISCUSSION TITLE
Testing

### DISCUSSION THREAD

**Discussion by ghost** [littleK0i/SnowDDL#75] [Q&A]
Hi. I am trying to run the `run_test.sh` after running the following on a clean account:

```
USE ROLE ACCOUNTADMIN;

CREATE ROLE SNOWDDL_ADMIN;

GRANT ROLE SYSADMIN TO ROLE SNOWDDL_ADMIN;
GRANT ROLE SECURITYADMIN TO ROLE SNOWDDL_ADMIN;

CREATE USER SNOWDDL
PASSWORD = 'REDACTED'
DEFAULT_ROLE = SNOWDDL_ADMIN;

GRANT ROLE SNOWDDL_ADMIN TO USER SNOWDDL;
GRANT ROLE SNOWDDL_ADMIN TO ROLE ACCOUNTADMIN;



CREATE WAREHOUSE SNOWDDL_WH
WAREHOUSE_SIZE = 'XSMALL'
AUTO_SUSPEND = 60
AUTO_RESUME = TRUE
INITIALLY_SUSPENDED = TRUE;

GRANT USAGE, OPERATE ON WAREHOUSE SNOWDDL_WH TO ROLE SNOWDDL_ADMIN;

ALTER USER SNOWDDL SET DEFAULT_WAREHOUSE = SNOWDDL_WH;
```

But the test script complains immediately with:

```
Granting role SNOWDDL__ACCOUNTADMIN to SNOWDDL_ADMIN would create a cycle., 'GRANT ROLE "SNOWDDL__ACCOUNTADMIN" TO ROLE "SNOWDDL_ADMIN"')
```

What's the right set up to run all the tests and see them pass? :) 

I want to extend the test suite to include STREAM.

**Comment by littleK0i** [comment]
Here is the script to set up Snowflake account for tests: https://github.com/littleK0i/SnowDDL/blob/master/test/_sql/account_setup.sql

Since we want to be able to actually run everything during testing, the SnowDDL admin role is derived from `ACCOUNTADMIN` instead of `SYSADMIN + SECURITYADMIN`.

So.. this would let you run tests, but it creates another problem. Normal `ACCOUNTADMIN` would not be able to see objects created by SnowDDL admin, since SnowDDL admin role is higher in the hierarchy.

In order to mitigate this, you may create an additional "super-admin" role for human user and grant SnowDDL role to it. With this approach you'll be able to see objects created by tests in Snowflake GUI.

Unfortunately, this is caused by fundamental flaws in Snowflake design, so I am currently not aware of any viable workarounds.

```sql
CREATE ROLE SUPER_ADMIN;
GRANT ROLE SUPER_ADMIN TO USER <your_admin_user>;
ALTER USER <your_admin_user> SET DEFAULT_ROLE = SUPER_ADMIN;

GRANT ROLE ACCOUNTADMIN TO ROLE SUPER_ADMIN;
GRANT ROLE SNOWDDL_ADMIN_TEST TO ROLE SUPER_ADMIN;
```

A few other tips about tests:

1. Make sure to use a dedicated Snowflake account for tests. Never run it on production or dev account, you may wipe all data accidentally.
2. Previously trial account worked just fine, but it is no longer sufficient. Snowflake introduced a lot of limitations for trial accounts recently. For example, EXTERNAL ACCESS INTEGRATION is not available on trial. So.. it should be a "paid" account.
3. Make sure to add a global account-wide RESOURCE_MONITOR with very low credit limit to prevent accidental overspend. A few credits per day should suffice.
